docs/proposals/serialization.md

```diff
@@ -0,0 +1,212 @@
+# Block retrieval optimization
+
+This document aims to provide a detailed overview of a certain
+development problem and suggest/compare several possible solutions. We
+will mostly talk about _speeding up_ data
+deserialization/verification and diffusion interface.
+
+## Terms and definitions
+
+Arbitrary datatype `T` construction/deserialization routine usually
+consists of the following steps:
+
+ * __(D1)__ Read bytes (from network or any type of storage)
+ * __(D2)__ Check CBOR structure (bytestring is a correct CBOR encoded data)
+ * __(D3)__ Check datatype canonicity
+ * __(D4)__ Construct/deserialize datatype `T`
+   * __(D4.1)__ An option is to deserialize only some part of data (header
+     of block)
+ * __(D5)__ Validate `T`.
+   * __(D5.1)__ Validate parts of `T`.
+
+Current codebase only provides interface to do all of it at
+once. User runs D1 and feeds bytestring to decoder. Decoder performs
+D2-D3, parses subfields and then constructs datatype (D4) with a
+special constructor (D5).
+
+It is currently impossible to perform (D4.1) and (D5.1), though it can
+be implemented.
+
+Now let's describe features that we'd like to have. We'll formulate
+them in terms of use cases:
+ * __(U1)__ Respond block requests quickly.
+     * Only D1 is needed (just get those bytes and send them).
+ * __(U2)__ Diffusion (as designed for Shelley) of blocks: receive block
+   and distribute it on fly.
+     * __(U2.1)__ D1, D2, D3 and D4.1 are needed. Regarding blocks, we don't want
+       to deserialize payload at all, checking header is enough.
+     * __(U2.2)__ One can argue that may want to have D4 + D5.1 instead of
+       D4.1, if it's not significantly slower.
+ * __(U3)__ Apply received block(s) to GState.
+     * D1-D5 (complete procedure) are needed.
+ * __(U4)__ Reading data (sometimes lots of it) from DB for different
+   node local actions (LRC, GState checks, API for Middleware -- all
+   of them require db traversing or LCA computations).
+     * D1,D4 needed (we trust local database, so we want to turn off
+       as much checks as it's possible).
+     * AFAIU @avieth assumes D2-D3 is also needed and we shouldn't
+       optimize here because it's too dangerous anyway.
```

Comment:
Yes D2 is certainly needed. How can we get the datatype structure if we don't use its CBOR encoding? As for D3, I suppose we could drop it, but honestly I don't think it's worth it.

Reply:
Agree. I will change it to `D1, D2, D4 needed`, ok? I personally think it will be a noticeable optimization since, for example, `decodeMapSkel` does `reverse` and also O(n) order checks. It's a pretty expensive computation for large `n`s -- of course it doesn't change the asymptotic (still linear), but reduces the constant (~3 times faster?).